KAFKA-16557 Implemented OffsetFetchRequestState toStringBase and added a test for it in CommitRequestManagerTest

brenden20 opened the pull request:
I changed OffsetFetchRequestState toString to toStringBase. I also added the instance fields from OffsetFetchRequestState and RetriableState to the toStringBase method. Within RetriableState, I also added a getter method for expirationTimeMs for testing purposes.

I added a new test to CommitRequestManagerTest that tests the new toStringBase method. In the test, a string variable 'target' is instantiated and will be compared to the new toStringBase method. The test calls OffsetFetchRequestState toStringBase() and compares it to the target string. All tests in CommitRequestManagerTest pass.

brenden20 committed:
Changed toString() override to toStringBase()

Changed toString() override to toStringBase() and implemented some fields in toStringBase()

brenden20 committed:
Added test for toStringBase()

Added test for OffsetFetchRequestState.toStringBase().

brenden20 committed:
Added to toStringBase()

Added more relevant fields in toStringBase() and updated the test to reflect the changes in OffsetFetchRequestState.toStringBase()

brenden20 committed:
Add println and example hashcode

Added println to toStringBase() test to see if it works. Also added example hashcode

brenden20 committed:
Edited toStringBase test

Added an assertEquals statement

brenden20 committed:
Added to toStringBase and updated test

Added fields inherited from RetriableRequestState to the toStringBase method. Had to add a get method for expirationTimeMs to make it work.

Also updated the test to reflect the changes.

brenden20 committed:
Remove hashcode comment

Remove hashcode comment

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:
```diff
@@ -227,8 +275,8 @@ public void testPollEnsureEmptyPendingRequestAfterPoll() {
         Map<TopicPartition, OffsetAndMetadata> offsets = Collections.singletonMap(
-            new TopicPartition("topic", 1),
-            new OffsetAndMetadata(0));
+                new TopicPartition("topic", 1),
```
could you undo these changes that aren't related to the pr?

brenden20 replied:
When I try to change it back, it says that the file is the same. I copied the parts of the test that I did not change from a different branch but Git is saying the files are the same still. Not too sure why it showed I made indentations, they are not showing up on my end in Intellij.

philipnee replied:
i wonder if intellij's auto indentation automatically changes the indentation when copying the content over.  double check the indentation for these files they should be there.  it could be easier to use a text editor like vim/emac to edit out these unintended changes.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:
```diff
@@ -880,6 +880,10 @@ boolean retryTimeoutExpired(long currentTimeMs) {
         }
 
         abstract void removeRequest();
+
```
add a comment // Visible for testing

brenden20 replied:
Got that added in a new commit!

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:
```diff
@@ -1090,14 +1094,15 @@ private void chainFuture(
                     ", future=" + future +
-                    ", " + toStringBase() +
-                    '}';
+                    ", memberId=" + memberInfo.memberId.orElse("undefined") +
+                    ", memberEpoch=" + (memberInfo.memberEpoch.isPresent() ? memberInfo.memberEpoch : "undefined");
```
My preference would be to add a `toString()` to the `MemberInfo` class. WDYT?

brenden20 replied:
Thank you for the suggestion, I added a toString() method to MemberInfo in my latest commit.

kirktrue reviewed (changes requested):
Thanks for the PR, @brenden20!

I'd never really thought to add a unit test for `toString()`, but it's not a bad idea 👍 

Just to echo @philipnee, revert any unnecessary whitespace/indentation changes.

brenden20 committed:
Add testing comment

Added visible for testing comment

brenden20 committed:
Updated toStringBase and added toString

Updated toStringBase() to have better output formatting, also added a toString() method for MemberInfo class.

Updated toStringBase() test to reflect changes

kirktrue commented:
@brenden20, as mentioned on another one of your PRs, there's a checkstyle violation here. You can run this command locally to avoid waiting for the CI infrastructure to catch it:

```
./gradlew check -x test
```

brenden20 committed:
Remove debugging print

Remove debugging print

brenden20 committed:
Merge branch 'apache:trunk' into 16557

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:
```diff
@@ -123,6 +126,47 @@ public void setup() {
+        final long retryBackoffMs = 10;
+        final long retryBackoffMaxMs = 100;
+        final long expirationTimeMs = 1000;
+        ConsumerConfig config = mock(ConsumerConfig.class);
```
try be consistent with the syntax.  either use final for the config and memberinfo or just not use final for the longs.

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:
```diff
@@ -123,6 +126,47 @@ public void setup() {
+
+        String target = requestState.toStringBase() +
+                ", memberInfo={" + offsetFetchRequestState.memberInfo +
+                "}, expirationTimeMs=" + (offsetFetchRequestState.expirationTimeMs().isPresent() ? offsetFetchRequestState.expirationTimeMs() : "undefined") +
```
i think we are expecting an expirationTimeMs so just use `expirationTimeMs=" + expirationTimeMs`

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:
```diff
@@ -123,6 +126,47 @@ public void setup() {
+                retryBackoffMaxMs);
+
+        String target = requestState.toStringBase() +
+                ", memberInfo={" + offsetFetchRequestState.memberInfo +
```
we can just test the values against L131-135, i think target shoud just be:
memberInfo={memberInfo}, as well as other fields below.

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:
```diff
@@ -123,6 +126,47 @@ public void setup() {
+
+        this.offsetFetchRequestState = commitRequestManager.new OffsetFetchRequestState(
+                mock(Set.class),
+                retryBackoffMs, retryBackoffMaxMs, expirationTimeMs,
```
let break this into separated line just to be consistent with L154-158.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:
```diff
@@ -1289,5 +1293,11 @@ static class MemberInfo {
         }
+
+        @Override
+        public String toString() {
```
Add the class name.  reformat it to
```
MemberInfo(memberId=..., memberEpoch=...)
```

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:
```diff
@@ -106,6 +106,9 @@ public class CommitRequestManagerTest {
     private OffsetCommitCallbackInvoker offsetCommitCallbackInvoker;
     private final Metrics metrics = new Metrics();
     private Properties props;
+    private CommitRequestManager.OffsetFetchRequestState offsetFetchRequestState;
```
the `commitRequestManager`, `offsetFetchRequestState` and `requestState` can all stay in the scope of `testOffsetFetchRequestStateToStringBase` no?

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:
```diff
@@ -123,6 +126,47 @@ public void setup() {
+                memberInfo
+        );
+
+        this.requestState = new RequestState(
```
this should be `new RetriableRequestState` since offsetFetchRequestState implements the retriable one?

philipnee commented:
hi @brenden20 - left some comments. lmk what do you think.

brenden20 committed:
Implementing PR suggestions

Updated testOffsetFetchRequestStateToStringBase(), changed some variable scope and changed the target string slightly. Also made a small update to toStringBase() and MemberInfo.toString()

brenden20 committed:
Fixed build error

Fixed build error

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:
```diff
@@ -123,6 +123,46 @@ public void setup() {
+                memberInfo
+        );
+
+        RequestState requestState = new RequestState(
```
we should use `RetriableRequestState` right? such that we can get the expirationTimeMs

brenden20 replied:
RetriableRequestState is abstract, I do not think that I can instantiate that class directly

philipnee replied:
you are right, scratch that.

brenden20 replied:
Gotcha, I will mark this suggestion as resolved then.

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:
```diff
@@ -123,6 +123,46 @@ public void setup() {
+                ", isExpired=" + offsetFetchRequestState.isExpired +
+                ", requestedPartitions=" + offsetFetchRequestState.requestedPartitions +
+                ", future=" + offsetFetchRequestState.future();
+
```
can we also assert no exception can be thrown for calling toString() - just to ensure no lousy toString().

brenden20 committed:
Small changes to testOffsetFetchRequestStateToStringBase()

Small changes to testOffsetFetchRequestStateToStringBase()

brenden20 committed:
Remove debug print

brenden20 committed:
Whitespace fixes

brenden20 committed:
More whitespace changes

brenden20 committed:
Fix style

Fix style

brenden20 committed:
Updated testOffsetFetchRequestStateToStringBase()

brenden20 committed:
Merge branch 'apache:trunk' into 16557

brenden20 committed:
Revert "Merge branch 'apache:trunk' into 16557"

This reverts commit f5abcc5bd956dc8157b2ff1e885edd8d72689cbd, reversing
changes made to 554ad5300620348e62bac210d1634755108ddc00.

kirktrue reviewed (approved):
LGTM! Can we move this from Draft to Ready to review?

brenden20 commented:
Marked ready to review

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:
```diff
@@ -123,6 +124,50 @@ public void setup() {
+
+        String target = requestState.toStringBase() +
+                ", memberInfo=" + memberInfo +
+                ", expirationTimeMs=" + (offsetFetchRequestState.expirationTimeMs().isPresent() ? offsetFetchRequestState.expirationTimeMs() : "undefined") +
```
we can directly do `expirationTimeMs=1000` here. in this case we do that we don't need to change the scope of `expirationTimeMs()`

brenden20 replied:
Looking at the code, I am able to do this, however it will cause an issue with OffsetFetchRequestState.toStringBase(). ```expirationTimeMs()``` is needed to implement the ```toStringBase()``` method, since ```OffsetFetchRequestState``` cannot access ```RetriableRequestState.expirationTimeMs``` since it is a private field.

brenden20 committed:
Updated testOffsetFetchRequestStateToStringBase()

Updated testOffsetFetchRequestStateToStringBase()

brenden20 committed:
Revert "Updated testOffsetFetchRequestStateToStringBase()"

This reverts commit 97c29c64d0e956381c7f71412c0b079ccf1b8303.

brenden20 committed:
Merge branch 'trunk' into 16557

brenden20 committed:
Revert "Merge branch 'trunk' into 16557"

This reverts commit c34f473eb850f4c2f9cb86839ccadf7471e0fcf0, reversing
changes made to df3e2aec6f29f25cdbe8e344cccd323268706d60.

brenden20 committed:
Merge branch 'apache:trunk' into 16557

brenden20 commented:
Had a lot of merge issues, was taking too much time to fix so I made a new branch with my changes. New PR is here: https://github.com/apache/kafka/pull/16291
